src/utils/transaction.ts

```diff
@@ -568,6 +570,137 @@ const transaction = {
     throw new ParseError('Invalid transaction version.');
   },
 
+  /**
+   * Convert a Transaction instance to the history object.
+   */
+  async convertTransactionToHistoryTx(
+    tx: Transaction | CreateTokenTransaction | NanoContract,
+    storage: IStorage
+  ): Promise<IHistoryTx> {
+    if (!tx.hash) {
+      throw new Error('To be a history tx a calculated hash is required');
+    }
+
+    const inputs: IHistoryInput[] = [];
+    const outputs: IHistoryOutput[] = [];
+
+    for (const input of tx.inputs) {
+      const spentTx = await storage.getTx(input.hash);
+      let spentOut: IHistoryOutput;
+
+      if (spentTx) {
+        if (input.index >= spentTx.outputs.length) {
+          throw new Error(
+            `Index (${input.index}) outside of transaction output array bounds (${spentTx.outputs.length})`
+          );
+        }
+        spentOut = spentTx.outputs[input.index];
+      } else {
+        // Get from API
+        spentOut = await new Promise((resolve, reject) => {
+          txApi
+            .getTransaction(input.hash, response => {
+              if (!response.success) {
+                return reject(new Error(response.message ?? ''));
+              }
+
+              if (input.index >= response.tx.outputs.length) {
+                return reject(new Error('Index outsite of tx output array bounds'));
+              }
+              const out = response.tx.outputs[input.index];
+              const outRet = {
+                value: out.value,
+                token_data: out.token_data,
+                script: out.script,
+                spent_by: out.spent_by,
+                decoded: {},
+              } as IHistoryOutput;
+              if (out.token) {
+                outRet.token = out.token;
+              }
+              if (out.decoded.address) {
+                outRet.decoded.type = (out.decoded.type as string).toLocaleLowerCase();
+                outRet.decoded.address = out.decoded.address as string;
+                outRet.decoded.timelock = out.decoded.timelock as number;
+              }
+              // XXX: should we cover nano and other types?
+
+              return resolve(outRet);
+            })
+            .catch(err => reject(err));
+        });
+
+        // If we could not get tx from api or any other error happened
+        if (!spentOut) {
+          throw new Error(`Could not fetch tx ${tx.hash} from fullnode`);
```

Comment:
issue: I think this is important to test, as it's handling errors for a pain point for both users and developers.

Reply:
Added unit tests for these conditions